### ISSUE TITLE
v2代码中为什么没有对prompt进行排序？

### ISSUE THREAD

**Issue by aliasGitee** [csguoh/MambaIR#115]
作者你好，v2代码中对x进行了重排序，但是为什么没有对prompt进行重排序？在mamba扫描的源码中，对C矩阵的操作在这里：
y = (hs @ C.unsqueeze(-1)).squeeze(3) # (B, L, ED, N) @ (B, L, N, 1) -> (B, L, ED, 1)
hs 是重排序后，经过mamba扫描的输出数据，此时，C的顺序应该与hs一致，但是在v2代码中，prompt在未重排序的情况下就与C相加，这是否存在问题？

这是mamba的源码：
def selective_scan_seq(self, x, delta, A, B, C, D):
        # x : (B, L, ED) == ori_x
        # Δ : (B, L, ED) == linear(ori_x, d_inner, dt_rank), linear(dt_rank, d_inner)
        # A : (ED, N) == - nn.Parameter(arange(1, d_state + 1, dtype=torch.float32).repeat(d_inner, 1)), [[1,2,...,8],[1,2,...,8],...]
        # B : (B, L, N) == linear(ori_x, d_inner, d_state)
        # C : (B, L, N) == linear(ori_x, d_inner, d_state)
        # D : (ED) == nn.Parameter(ones(d_inner))

        # y : (B, L, ED)
        # ED: d_inner
        # N: d_state

        _, L, _ = x.shape

        deltaA = torch.exp(delta.unsqueeze(-1) * A) # (B, L, ED, 1) * (ED, N) => (B, L, ED, N)
        deltaB = delta.unsqueeze(-1) * B.unsqueeze(2) # (B, L, ED, 1) * (B, L, 1, N) => (B, L, ED, N)

        BX = deltaB * (x.unsqueeze(-1)) # (B, L, ED, N) * (B, L, ED, 1) => (B, L, ED, N)

        h = torch.zeros(x.size(0), self.config.d_inner, self.config.d_state, device=deltaA.device) # (B, ED, N)
        hs = []

        for t in range(0, L):
            h = deltaA[:, t] * h + BX[:, t]
            hs.append(h)

        hs = torch.stack(hs, dim=1) # (B, L, ED, N)

        y = (hs @ C.unsqueeze(-1)).squeeze(3) # (B, L, ED, N) @ (B, L, N, 1) -> (B, L, ED, 1)

        y = y + D * x

        return y

**Comment by ordinary7Zz** [comment]
你好，我在阅读这里的代码的时候也产生了同样的问题，请问你现在理解原因了吗？还是说确实是需要把prompt也进行重排呢？

**Comment by aliasGitee** [comment]
> 你好，我在阅读这里的代码的时候也产生了同样的问题，请问你现在理解原因了吗？还是说确实是需要把prompt也进行重排呢？

从代码上看应该是需要重排的，否则prompt就没意义了

**Comment by WilliamForest233** [comment]
Hi,
I was looking into the same question and made the following modifications to the code:
```python
semantic_x = semantic_neighbor(x, x_sort_indices)

semantic_prompt = semantic_neighbor(prompt, x_sort_indices)
y = self.selectiveScan(semantic_x, semantic_prompt)

y = self.out_proj(self.out_norm(y))
x = semantic_neighbor(y, x_sort_indices_reverse)
```
With these changes, the model's performance on my task turned out to be _slightly better_.

**Comment by csguoh** [comment]
Hi, 
I would like to clarify that the prompt is not re-odered since we want it to query across the image. For instance, if we un-re-order the i-th prompt, this prompt can potentially see any pixels in the image due to our re-order. However, if we also respectively re-order the prompt, the i-th prompt could still only see the i-th pixel, which constrains its receptive field.
